File: 4-How to solve coding problems/htscp/lib/contains_common_item.ex
defmodule ContainsCommonItem do
  def items(lst1, lst2) do
    # Loop through first list and create map where properties === items in
    # the list
    map = %{}
    map = Enum.reduce(lst1, map, &Map.put(&2, &1, true))

    # loop through second list and check if item in second list exists on
    # created map
    Enum.any?(lst2, &Map.has_key?(map, &1))

    # map =
    #   Enum.reduce(lst1, map, fn x, acc ->
    #     Map.put(acc, x, 1)
    #   end)

    # Enum.any?(lst2, fn x -> Map.has_key?(map, x) end)
  end

  # def items2(lst1, lst2) do
  #   not MapSet.disjoint?(MapSet.new(lst1), MapSet.new(lst2))
  # end
end

# O(a + b) Time Complexity
# ContainsCommonItem.items(["a", "b", "c", "x"], ["z", "y", "a"])
# ContainsCommonItem.items2(["a", "b", "c", "x"], ["z", "y", "a"])

File: 4-How to solve coding problems/contains_common_item.rb
# frozen_string_literal: true

def contains_common_item(arr1, arr2)
  # Loop through first array and create object where properties === items in
  # the array
  h = {}

  arr1.each do |i|
    h[i] = true unless h.include?(i)
  end

  # loop through second array and check if item in second array exists on
  # created object
  arr2.each do |j|
    return true if h.include?(j)
  end

  false
end

# O(a + b) Time Complexity
array1 = %w[a b c x]
array2 = %w[z y a]

puts contains_common_item(array1, array2)

# def contains_common_item2(arr1, arr2)
#   return true if arr1.intersect?(arr2)

#   false
# end

# puts contains_common_item2(array1, array2)
